school/database.py

```diff
@@ -0,0 +1,514 @@
+            .filter_by(guild_id=ctx.guild.id)
+            .filter_by(abbreviation=abbreviation.upper())
+        )
+
+        return query.one_or_none()
+
+    @staticmethod
+    def from_json(ctx, json_data):
+        abbreviation = json_data.get("abbreviation", None)
+
+        if not abbreviation:
+            return None
+
+        subject = (
+            session.query(Subject)
+            .filter_by(abbreviation=abbreviation, guild_id=ctx.guild.id)
+            .one_or_none()
+        )
+
+        if not subject:
+            subject = Subject(
+                guild_id=ctx.guild.id,
+                abbreviation=abbreviation,
+            )
+
+        name = json_data.get("name", None)
+        institute = json_data.get("institute", None)
+        winter_semester = json_data.get("winter_semester", False)
+        summer_semester = json_data.get("summer_semester", False)
+
+        if winter_semester and summer_semester:
+            semester = "Both"
+        elif winter_semester:
+            semester = "Winter"
+        else:
+            semester = "Summer"
+
+        subject.name = name
+        subject.institute = institute
+        subject.semester = semester
+
+        session.merge(subject)
+        session.flush()
+
+        query = session.query(SubjectUrl).filter_by(subject_idx=subject.idx).all()
+        for relation in query:
+            session.delete(relation)
+
+        json_url = json_data.get("link", None)
+
+        for url_str in json_url:
+            subject.url.append(SubjectUrl(subject_id=subject.idx, url=url_str))
+
+        session.commit()
+
+        return subject
+
+    def edit(self, abbreviation, name, institute, semester, guarantor):
+        if abbreviation:
+            self.abbreviation = abbreviation
+        if name:
+            self.name = name
+        if institute:
+            self.institute = institute
+        if semester:
+            self.semester = semester
+        if guarantor:
+            self.guarantor = guarantor
+
+        session.commit()
+
+    def set_guarantor(self, guarantor: Teacher):
+        self.guarantor = guarantor[0] if len(guarantor) != 0 else None
+        session.commit()
+
+    def set_teachers(self, teachers: List[Teacher]):
+        self.teachers = teachers
+        session.commit()
+
+    def add_teachers(self, teachers: List[Teacher]) -> List[str]:
+        ignored = []
+        for teacher in teachers:
+            if teacher in self.teachers:
+                ignored.append(str(teacher.school_id))
+            else:
+                self.teachers.append(teacher)
+
+        session.commit()
+
+        return ignored
+
+    def remove_teachers(self, teachers: List[Teacher]) -> List[str]:
+        ignored = []
+        for teacher in teachers:
+            if teacher not in self.teachers:
+                ignored.append(str(teacher.school_id))
+            else:
+                self.teachers.remove(teacher)
+
+        session.commit()
+
+        return ignored
```

Comment:
This is about the possibility to edit the data manually, so you don't have to import all data the data. Because the editign is not as easy as other attributes, I'd like to have this here as function instead in code and using save().

Reply:
Alright but in that case I wouldn't it make more sense to return the removed instead of the `not removed` teachers? Same goes for the added ones